azure/hostedservice/client.go

```diff
@@ -0,0 +1,135 @@
+package hostedservice
+
+import (
+	"encoding/base64"
+	"encoding/xml"
+	"fmt"
+
+	"github.com/MSOpenTech/azure-sdk-for-go/azure"
+	locationclient "github.com/MSOpenTech/azure-sdk-for-go/azure/location"
+)
+
+const (
+	azureXmlns                        = "http://schemas.microsoft.com/windowsazure"
+	azureDeploymentListURL            = "services/hostedservices/%s/deployments"
+	azureHostedServiceListURL         = "services/hostedservices"
+	deleteAzureHostedServiceURL       = "services/hostedservices/%s?comp=media"
+	azureHostedServiceAvailabilityURL = "services/hostedservices/operations/isavailable/%s"
+	azureDeploymentURL                = "services/hostedservices/%s/deployments/%s"
+	deleteAzureDeploymentURL          = "services/hostedservices/%s/deployments/%s?comp=media"
+
+	errParamNotSpecified = "Parameter %s is not specified."
+	errInvalidDnsLength  = "The DNS name must be between 3 and 25 characters."
+)
+
+//NewClient is used to return a handle to the HostedService API
+func NewClient(client *azure.Client) *HostedServiceClient {
+	return &HostedServiceClient{client: client}
+}
+
+func (self *HostedServiceClient) CreateHostedService(dnsName, location string, reverseDnsFqdn string, label string, description string) (string, error) {
+	if len(dnsName) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "dnsName")
+	}
+	if len(location) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "location")
+	}
+
+	err := self.verifyDNSName(dnsName)
+	if err != nil {
+		return "", err
+	}
+
+	result, reason, err := self.CheckHostedServiceNameAvailability(dnsName)
+	if err != nil {
+		return "", err
+	}
+	if !result {
+		return "", fmt.Errorf("%s Hosted service name: %s", reason, dnsName)
+	}
+
+	locationClient := locationclient.NewClient(self.client)
```

Comment:
You should defer this error check to REST API. This is special logic that shouldn't be on the lib. API will already tell you if the location is invalid. Tomorrow if API accepts some code other than location 's "name" then the lib will be broken in a way that user can't fix. We should not do this check.